OlympusDAO-Agent-Suit/TotalReserves/src/reserve.fetcher.ts

```diff
@@ -0,0 +1,32 @@
+import abi from "./abi";
+import LRU from "lru-cache";
+import { BigNumber, Contract, providers } from "ethers";
+
+export default class ReserveFetcher {
+  private cache: LRU<number, BigNumber>;
+  readonly treasury: string;
+  private tContract: Contract;
+
+  constructor(contract: string, provider: providers.Provider){
+    this.treasury = contract;
+    this.tContract = new Contract(contract, abi.TREASURY_ABI, provider);
+    this.cache = new LRU<number, BigNumber>();
+  }
+
+  public set(block: number, reserve: BigNumber) {
+    this.cache.set(block, reserve);
+  }
+
+  public async getReserve(block: number): Promise<BigNumber> {
+    if(this.cache.get(block) !== undefined)
+      return this.cache.get(block) as BigNumber;
+
+    const reserve = await this.tContract.totalReserves({ blockTag: block - 1 });
```

Comment:
Why is `blockTag` `block - 1`?

Reply:
to fetch the `totalReserves` at the very beginning of the current block, i.e the `totalReserves` at the end of the previous block.

Reply:
Great, but this should be handled when calling the method. I mean, you should do `getReserve(block - 1)`, doing `getReserve(block)` and getting reserve for `block - 1` is a little confusing.